bt-core/src/main/java/bt/net/ConnectionSource.java

```diff
@@ -110,72 +108,109 @@ public ConnectionResult getConnection(Peer peer, TorrentId torrentId) {
+            }
+            if (throwable != null) {
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug("Failed to establish outgoing connection to peer: " + peer, throwable);
                 }
-            });
+            }
+        });
+    }
 
-            pendingConnections.put(key, connection);
+    /**
+     * Checks whether a connection to this peer on the specified torrent id is possible. Returns a result if a new
+     * connection is not possible. This can happen if there is an existing pending connection, or we have reached
+     * {@link Config#getMaxPeerConnections()}
+     *
+     * @param peer      the peer to connect to
+     * @param torrentId the torrent for the connection
+     * @param key       the connection key
+     * @return a result if a new connection is not possible, null otherwise
+     */
+    private CompletableFuture<ConnectionResult> validateNewConnPossible(Peer peer, TorrentId torrentId,
+                                                                        ConnectionKey key) {
+        CompletableFuture<ConnectionResult> connection = getExistingOrPendingConnection(key);
+        if (connection != null) {
+            if (connection.isDone() && LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Returning existing connection for peer: {}. Torrent: {}", peer, torrentId);
+            }
             return connection;
         }
+
+        if (checkPeerConnectionsLimit(peer, torrentId)) {
+            return CompletableFuture.completedFuture(ConnectionResult.failure("Connections limit exceeded"));
+        }
+        return null;
     }
 
-    private CompletableFuture<ConnectionResult> getExistingOrPendingConnection(ConnectionKey key) {
-        PeerConnection existingConnection = connectionPool.getConnection(key);
-        if (existingConnection != null) {
-            return CompletableFuture.completedFuture(ConnectionResult.success(existingConnection));
+    private boolean checkPeerConnectionsLimit(Peer peer, TorrentId torrentId) {
+        if (connectionPool.size() >= config.getMaxPeerConnections()) {
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Will not attempt to establish connection to peer: {}. " +
+                        "Reason: connections limit exceeded. Torrent: {}", peer, torrentId);
+            }
+            return true;
         }
+        return false;
+    }
+
+    private CompletableFuture<ConnectionResult> getExistingOrPendingConnection(ConnectionKey key) {
+        // The order of checking these maps is crucial to avoid syncronization. Check pending connections, then
+        // existing connections. If the key is being moved from pendingConnections to existing connections, either
+        // We'll get it before it is removed from pendingConnections, or if it was removed, it must already be in
+        // existingConnections because ConcurrentMaps are guaranteed to have "happens-before" memory consistency
```

Comment:
I'm a bit uncomfortable with the assertiveness of this comment given that `connectionPool` variable's type is an interface :)

Reply:
I agree - the comment incorrectly implies that`IPeerConnectionPool.getConnection()` must be a ConcurrentMap. I will change it.

I believe that the issue is only related to the comment, and not the code.

The implementation of `IPeerConnectionPool` isn't assumed. `pendingConnections.remove()` is a remove operation in a concurrent collection and creates is a memory barrier. Therefore if we observe `pendingConnections.remove()` was called in another thread, `connectionPool.addConnectionIfAbsent()` must have already been called, making the connection visible in `connectionPool`.

Reply:
Yeah, I was only talking about the assumption that IPeerConnectionPool.getConnection is backed by a concurrent map. The code is alright.

Reply:
On closer analysis, this code is wrong. remove() does not have a happens before guarantee in the documentation. See https://stackoverflow.com/questions/39341742/does-concurrentmap-remove-provide-a-happens-before-edge-before-get-returns-n

Reply:
Made a new MR.